### ISSUE TITLE
Server lag (it's not the client)

### ISSUE THREAD

**Issue by aerozoic** [MinetestForFun/server-minetestforfun-skyblock#154]
I have confirmed the lag is not on the client side. This is server lag. Right now there are 5 players online and they all say it's lagging very badly. Max_Lag reports 53 seconds. This may be abnormally high at the moment, but even when the server is operating normally there is still 1-5 seconds of lag. That 1-5 seconds is what i have actually experienced personally at other times in the game, it is NOT from max_lag reports.
:large_orange_diamond:

**Comment by BetterToAutomateTheWorld** [comment]
Yes, it was a DDOS (again), it's laggy the time mitigation do its job...

**Comment by aerozoic** [comment]
I don't mean to be a dick (plz don't kill the messenger) but this isn't fixed. It was lagging all day today. Sometimes as much as 15 seconds. Again, this was my actual in game lag. This was happening with only 8 players online. I also saw several other players reporting lag throughout the day.
:large_orange_diamond:

**Comment by BetterToAutomateTheWorld** [comment]
Strange, I will check this again

But every times I checked it, i didn't see a "big" lag as you describe it

**Comment by aerozoic** [comment]
It's lagging pretty bad right now.
:large_orange_diamond:

**Comment by aerozoic** [comment]
I just now checked again. Perhaps there's another DDOS? It was lagging HORRIBLY. I kept getting connection timed out errors and it took several minutes before i could even join the server. Once i was in, the lag was just terrible. I tried to TP to somebody and after the TP i got stuck because it took nearly a minute for the map to load. I also tried opening doors and transferring items to/from a chest which also lagged.
:large_orange_diamond:

**Comment by BetterToAutomateTheWorld** [comment]
I niced the server processes https://github.com/MinetestForFun/xorhub-serverscontents/commit/114577d8a82e6b7839f5893f4d6913ccc3d6c636
We need to check tomorrow if it changes something or not

**Comment by BetterToAutomateTheWorld** [comment]
Ok, I did again some tweaks, we need to check again the lag on SB after the 5th of this month

**Comment by BetterToAutomateTheWorld** [comment]
Up, @cessna151 are lags still here ? The server now runs really differently 
![cpu_ram_before_after](http://i.imgur.com/AjFm77h.png)

CPU : won't exceed 50% now
RAM : Won'texceed 6~7GB now

**Comment by aerozoic** [comment]
The lag has reduced greatly!  Now it seems there is a short pause (lag) of 1-2 seconds every time a player joins or leaves.
:large_orange_diamond:

**Comment by BetterToAutomateTheWorld** [comment]
For the joining of new players, the microlag is something needed (new island schematic created), but for the leaving of a player, that's really strange... We need investigation

**Comment by BetterToAutomateTheWorld** [comment]
Since now, LuaJIT is enabled on the SKYBLOCK server, can you tell us again (i know, sorry :p) if you see lag on the server ? Thank you in advance.

**Comment by aerozoic** [comment]
A little lag here and there today, then tonight the lag just got horrible! :(
:large_orange_diamond:

**Comment by aerozoic** [comment]
After a couple days of playing i can say that, in general, there is still a lag problem.
:large_orange_diamond:

**Comment by ghost** [comment]
WTF skyblock:round_down??  https://github.com/MinetestForFun/server-minetestforfun-skyblock/commit/982fcbf2f4f441f709bcdb04673598eb2f401b57

I make 30000  entry in skyblock.spawn (my player at the middle) and i placed  mores skyblock:round_down.
lag 0.1 before,  0.6 after, if more players place this bad node then the lag grow.

skyblock.get_spawn_player() https://github.com/MinetestForFun/server-minetestforfun-skyblock/blob/master/mods/skyblock/skyblock.lua#L96

until 9.0s lag if my player at the end of skyblock.spawn.

**Comment by BetterToAutomateTheWorld** [comment]
![high_priority](http://i.imgur.com/nPh6AQb.png)

Hi,

## Discovery

After more explanations from @crabman77 and a discussion with him, we point out the big problem isn't really **skyblock:round_down** but the protector mod.

## How protector blocks work

First of all, how the protector blocks works ? 
So, yes they grant the possibility to players to protect some areas. Since the protector block mod is activated on the server, when a player did an action like dig a node or put a node, the following question is called from the server **Can this player dig/place this node ?** which result with the process(simplified for protector block purpose) :
- Is the player in a protected area ? yes/no
- If no, do we have a protector block near here ? Check the 16*16(= 256 nodes) area around the player position
- If a protector_block is in this area, what is the player name ? player_name
- Who is the owner of the protector_block ? player_name / other player
- Ok, you can dig here / No, you can't dig here
  => Summarized by the **minetest.find_nodes_in_area()** Minetest call

## History

Historically, the first protector block was a 16x16 protector block which can protect an area up to 256 nodes, but we introduce the 32x32 protector block... This block force the **minetest.find_nodes_in_area()** Minetest call to check 1024 nodes at each digged/placed nodes per player.

## Aggravating factors

The main problem today about the lags is the addition of :
- Mods like **skyblock:round_down** which call minetest.find_nodes_in_area()
- The **high number of placed protector blocks**
- The **number of digged/placed blocks** (of any kind)
- The **number of players**
- The (32x32=)1024 nodes checked every player actions

_Linked @LeMagnesium to this issue
Assigned @crabman77 @LeMagnesium @Darcidride to this issue_

**Comment by aerozoic** [comment]
It is still lagging :(
:large_orange_diamond:

**Comment by aerozoic** [comment]
Has there been a commit for this that is waiting to be merged? What's the latest on this issue?
:large_orange_diamond:
